hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/processors/StreamToStreamJoinP.java

```diff
@@ -0,0 +1,177 @@
+import com.hazelcast.sql.impl.expression.ExpressionEvalContext;
+import com.hazelcast.sql.impl.row.JetSqlRow;
+
+import javax.annotation.Nonnull;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.PriorityQueue;
+
+import static java.util.Collections.emptyList;
+import static java.util.Comparator.comparingLong;
+
+public class StreamToStreamJoinP extends AbstractProcessor {
+    // TODO: replace with ExpressionEvalContext.from(context) in init() method.
+    private static final ExpressionEvalContext MOCK_EEC =
+            new ExpressionEvalContext(emptyList(), new DefaultSerializationServiceBuilder().build());
+    /**
+     * <p>
+     * JOIN condition should be transformed into such form:
+     * <pre>
+     *  l.time >= r.time - constant1
+     *  r.time >= l.time - constant2
+     * </pre>
+     */
+    private final JetJoinInfo joinInfo;
+    private final ToLongFunctionEx<JetSqlRow> leftTimestampExtractor;
+    private final ToLongFunctionEx<JetSqlRow> rightTimestampExtractor;
+    private final Map<Byte, Long>[] postponeTimeMap;
+
+    private final PriorityQueue<JetSqlRow> leftBuffer;
+    private final PriorityQueue<JetSqlRow> rightBuffer;
+
+    private Iterator<JetSqlRow> leftPos;
+    private Iterator<JetSqlRow> rightPos;
+
+    private JetSqlRow leftItem;
+    private JetSqlRow rightItem;
+
+    private JetSqlRow pendingLeftOutput;
+    private JetSqlRow pendingRightOutput;
+    private Watermark pendingWatermark;
+
+    public StreamToStreamJoinP(
+            final JetJoinInfo joinInfo,
+            final ToLongFunctionEx<JetSqlRow> leftTimestampExtractor,
+            final ToLongFunctionEx<JetSqlRow> rightTimestampExtractor,
+            final Map<Byte, Long>[] postponeTimeMap
+    ) {
+        this.joinInfo = joinInfo;
+        this.leftTimestampExtractor = leftTimestampExtractor;
+        this.rightTimestampExtractor = rightTimestampExtractor;
+        this.postponeTimeMap = postponeTimeMap;
+
+        leftBuffer = new PriorityQueue<>(comparingLong(leftTimestampExtractor));
+        rightBuffer = new PriorityQueue<>(comparingLong(rightTimestampExtractor));
+    }
+
+    @Override
+    protected boolean tryProcess0(@Nonnull Object item) {
+        if (pendingLeftOutput != null) {
+            if (tryEmit(pendingLeftOutput)) {
+                pendingLeftOutput = null;
+            }
+            return false;
+        }
+
+        // left input, traverse right buffer
+        if (leftItem == null) {
+            leftItem = (JetSqlRow) item;
+            leftBuffer.offer(leftItem);
+            rightPos = rightBuffer.iterator();
+        }
+
+        if (!rightPos.hasNext()) {
+            leftItem = null;
+            return true;
+        }
+
+        pendingLeftOutput = ExpressionUtil.join(leftItem, rightPos.next(), joinInfo.condition(), MOCK_EEC);
+        if (pendingLeftOutput != null && tryEmit(pendingLeftOutput)) {
+            pendingLeftOutput = null;
+        }
+        return false;
```

Comment:
We should emit as many items as we can in one call:
```suggestion
        while (rightPos.hasNext()) {
            pendingLeftOutput = ExpressionUtil.join(leftItem, rightPos.next(), joinInfo.condition(), MOCK_EEC);
            if (pendingLeftOutput != null) {
                if (tryEmit(pendingLeftOutput)) {
                    pendingLeftOutput = null;
                } else {
                    return false;
                }
            }
        }
```